File: loxide/src/chunk.rs
use std::ops::Deref;

use crate::value::{Value, ValueArray};

#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
pub enum Opcode {
    Return = 0,
    Constant,
    Negate,
    Add,
    Subtract,
    Multiply,
    Divide,
    Nil,
    True,
    False,
    Not,
    Equal,
    Greater,
    Less,
    Print,
    Pop,
    DefineGlobal,
    GetGlobal,
    SetGlobal,
    GetLocal,
    SetLocal,
    JumpIfFalse,
    Jump,
    Loop,
    Call,
    Closure,
    GetUpvalue,
    SetUpvalue,
}

impl Opcode {
    pub fn from_u8(val: u8) -> Option<Self> {
        use Opcode::*;
        match val {
            0 => Some(Return),
            1 => Some(Constant),
            2 => Some(Negate),
            3 => Some(Add),
            4 => Some(Subtract),
            5 => Some(Multiply),
            6 => Some(Divide),
            7 => Some(Nil),
            8 => Some(True),
            9 => Some(False),
            10 => Some(Not),
            11 => Some(Equal),
            12 => Some(Greater),
            13 => Some(Less),
            14 => Some(Print),
            15 => Some(Pop),
            16 => Some(DefineGlobal),
            17 => Some(GetGlobal),
            18 => Some(SetGlobal),
            19 => Some(GetLocal),
            20 => Some(SetLocal),
            21 => Some(JumpIfFalse),
            22 => Some(Jump),
            23 => Some(Loop),
            24 => Some(Call),
            25 => Some(Closure),
            26 => Some(GetUpvalue),
            27 => Some(SetUpvalue),
            _ => None,
        }
    }
}

pub struct Chunk {
    pub code: Vec<u8>,
    pub constants: ValueArray,
    pub lines: Vec<u32>,
}

impl Chunk {
    pub fn new() -> Self {
        Self {
            code: vec![],
            constants: vec![],
            lines: vec![],
        }
    }

    pub fn iter(&self) -> ChunkIter {
        ChunkIter {
            chunk: self,
            offset: 0,
        }
    }

    pub fn iter_debug(&self) -> ChunkIterDebug {
        ChunkIterDebug(ChunkIter {
            chunk: self,
            offset: 0,
        })
    }

    pub fn write(&mut self, op: u8, line: u32) {
        self.code.push(op);
        self.lines.push(line);
    }

    pub fn add_constant(&mut self, value: Value) -> u8 {
        let index = self.constants.len();
        self.constants.push(value);
        index.try_into().unwrap()
    }

    /// Dissamble instruction and increment offset to the start of
    /// the next one
    pub fn disassemble_instruction(&self, offset: &mut usize) -> Option<Instruction> {
        let instr = self.code[*offset];
        let op = Opcode::from_u8(instr);
        match op {
            Some(
                Opcode::Pop
                | Opcode::Print
                | Opcode::Equal
                | Opcode::Greater
                | Opcode::Less
                | Opcode::Not
                | Opcode::True
                | Opcode::False
                | Opcode::Nil
                | Opcode::Add
                | Opcode::Subtract
                | Opcode::Multiply
                | Opcode::Divide
                | Opcode::Negate
                | Opcode::Return,
            ) => {
                *offset += 1;
                Some(Instruction::Simple(op.unwrap()))
            }
            Some(
                Opcode::Constant | Opcode::DefineGlobal | Opcode::GetGlobal | Opcode::SetGlobal,
            ) => {
                let constant_idx = self.code[*offset + 1];
                let constant = self.constants[constant_idx as usize];
                *offset += 2;
                Some(Instruction::Constant(op.unwrap(), constant))
            }
            Some(Opcode::GetLocal | Opcode::SetLocal | Opcode::Call) => {
                let slot = self.code[*offset + 1];
                *offset += 2;
                Some(Instruction::Byte(op.unwrap(), slot))
            }
            Some(Opcode::Jump | Opcode::JumpIfFalse | Opcode::Loop) => {
                let byte1 = self.code[*offset + 1];
                let byte2 = self.code[*offset + 2];
                *offset += 3;
                let val = ((byte1 as u16) << 8) | (byte2 as u16);
                Some(Instruction::Jump(op.unwrap(), val))
            }
            Some(Opcode::Closure) => {
                *offset += 1;
                let constant_idx = self.code[*offset];
                *offset += 1;
                let value = self.constants[constant_idx as usize];
                Some(Instruction::Closure(value))
            }
            otherwise => panic!("Invalid opcode {:?}", otherwise),
        }
    }
}

impl Deref for Chunk {
    type Target = Vec<u8>;

    fn deref(&self) -> &Self::Target {
        &self.code
    }
}

impl std::fmt::Debug for Chunk {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_list().entries(self.iter_debug()).finish()
    }
}

#[derive(Debug)]
pub struct InstructionDebug {
    pub line: u32,
    pub inner: Instruction,
}

pub enum Instruction {
    Simple(Opcode),
    Constant(Opcode, Value),
    Byte(Opcode, u8),
    Jump(Opcode, u16),
    Closure(Value),
}

impl std::fmt::Debug for Instruction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Instruction::Simple(op) => {
                write!(f, "{:?}", op)
            }
            Instruction::Constant(op, val) => {
                f.debug_tuple("Constant").field(op).field(val).finish()
            }
            Instruction::Byte(op, val) => f.debug_tuple("Byte").field(op).field(val).finish(),
            Instruction::Jump(op, offset) => f.debug_tuple("Jump").field(op).field(offset).finish(),
            Instruction::Closure(val) => f.debug_tuple("Closure").field(val).finish(),
        }
    }
}

pub struct ChunkIter<'a> {
    chunk: &'a Chunk,
    offset: usize,
}

impl<'a> ChunkIter<'a> {
    /// Dissamble instruction and increment offset to the start of
    /// the next one
    fn disassemble_instruction(&mut self) -> Option<Instruction> {
        self.chunk.disassemble_instruction(&mut self.offset)
    }
}

impl<'a> Iterator for ChunkIter<'a> {
    type Item = Instruction;

    fn next(&mut self) -> Option<Self::Item> {
        if self.offset >= self.chunk.len() {
            return None;
        }

        self.disassemble_instruction()
    }
}

pub struct ChunkIterDebug<'a>(ChunkIter<'a>);

impl<'a> Iterator for ChunkIterDebug<'a> {
    type Item = InstructionDebug;

    fn next(&mut self) -> Option<Self::Item> {
        if self.0.offset >= self.0.chunk.len() {
            return None;
        }

        let line = self.0.chunk.lines[self.0.offset];
        let inner = self.0.disassemble_instruction();

        inner.map(|inner| InstructionDebug { inner, line })
    }
}

File: zlox/src/chunk.zig
const Value = @import("value.zig").Value;
const std = @import("std");
const Obj = @import("obj.zig");
const mem = std.mem;
const debug = std.debug;
const Allocator = mem.Allocator;

const ArrayList = std.ArrayListUnmanaged;

pub const Opcode = enum(u8) { Return, Constant, Negate, Add, Subtract, Multiply, Divide, Nil, True, False, Not, Equal, Greater, Less, Print, Pop, DefineGlobal, GetGlobal, SetGlobal, GetLocal, SetLocal, JumpIfFalse, Jump, Loop, Call, Closure, GetUpvalue, SetUpvalue, CloseUpvalue, Class, GetProperty, SetProperty, Method, Invoke };

pub const Chunk = struct {
    const Self = @This();

    code: ArrayList(u8),
    constants: ArrayList(Value),
    lines: ArrayList(u16),

    pub fn init(allocator: Allocator) Allocator.Error!Self {
        return Chunk{
            .code = try ArrayList(u8).initCapacity(allocator, 64),
            .constants = try ArrayList(Value).initCapacity(allocator, 64),
            .lines = try ArrayList(u16).initCapacity(allocator, 4),
        };
    }

    pub fn write_op(self: *Chunk, allocator: Allocator, op: Opcode, line: u16) !void {
        try self.write_byte(allocator, @enumToInt(op), line);
    }

    pub fn write_byte(self: *Chunk, allocator: Allocator, byte: u8, line: u16) !void {
        try self.code.append(allocator, byte);
        try self.lines.append(allocator, line);
    }

    pub fn write_u16(self: *Chunk, allocator: Allocator, val: u16, line: u16) !void {
        try self.write_byte(allocator, @intCast(u8, val >> 8), line);
        try self.write_byte(allocator, @intCast(u8, val), line);
    }

    pub fn free(self: *Chunk, allocator: Allocator) void {
        self.code.clearAndFree(allocator);
        self.constants.clearAndFree(allocator);
        self.lines.clearAndFree(allocator);
    }

    pub fn add_constant(self: *Chunk, allocator: Allocator, value: Value) !u8 {
        const index = self.constants.items.len;
        try self.constants.append(allocator, value);
        return @intCast(u8, index);
    }

    pub fn disassemble(self: *const Chunk, name: []const u8) void {
        debug.print("== {s} ==\n", .{name});
        var offset: usize = 0;
        while (offset < self.code.items.len) {
            offset = self.disassemble_instruction(offset);
        }
        debug.print("== end {s} ==\n", .{name});
    }

    pub fn disassemble_instruction(self: *const Chunk, offset: usize) usize {
        debug.print("{d:04} ", .{offset});
        if (offset > 0 and self.lines.items[offset] == self.lines.items[offset - 1]) {
            debug.print("   | ", .{});
        } else {
            debug.print("{d:4} ", .{self.lines.items[offset]});
        }

        const instruction: Opcode = @intToEnum(Opcode, self.code.items[offset]);
        switch (instruction) {
            .Equal => {
                return simple_instruction("OP_EQUAL", offset);
            },
            .Greater => {
                return simple_instruction("OP_GREATER", offset);
            },
            .Less => {
                return simple_instruction("OP_LESS", offset);
            },
            .Not => {
                return simple_instruction("OP_NOT", offset);
            },
            .Nil => {
                return simple_instruction("OP_NIL", offset);
            },
            .True => {
                return simple_instruction("OP_TRUE", offset);
            },
            .False => {
                return simple_instruction("OP_FALSE", offset);
            },
            .Add => {
                return simple_instruction("ADD", offset);
            },
            .Subtract => {
                return simple_instruction("SUBTRACT", offset);
            },
            .Multiply => {
                return simple_instruction("MULTIPLY", offset);
            },
            .Divide => {
                return simple_instruction("DIVIDE", offset);
            },
            .Negate => {
                return simple_instruction("OP_NEGATE", offset);
            },
            .Return => {
                return simple_instruction("OP_RETURN", offset);
            },
            .Constant => {
                return self.constant_instruction("OP_CONSTANT", offset);
            },
            .Print => {
                return simple_instruction("OP_PRINT", offset);
            },
            .Pop => {
                return simple_instruction("OP_POP", offset);
            },
            .DefineGlobal => {
                return self.constant_instruction("OP_DEFINE_GLOBAL", offset);
            },
            .GetGlobal => {
                return self.constant_instruction("OP_GET_GLOBAL", offset);
            },
            .SetGlobal => {
                return self.constant_instruction("OP_SET_GLOBAL", offset);
            },
            .GetLocal => {
                return self.byte_instruction("OP_GET_LOCAL", offset);
            },
            .SetLocal => {
                return self.byte_instruction("OP_SET_LOCAL", offset);
            },
            .Jump => {
                return self.jump_instruction("OP_JUMP", 1, offset);
            },
            .JumpIfFalse => {
                return self.jump_instruction("OP_JUMP_IF_FALSE", 1, offset);
            },
            .Loop => {
                return self.jump_instruction("OP_LOOP", -1, offset);
            },
            .Call => {
                return self.byte_instruction("OP_CALL", offset);
            },
            .Closure => {
                var ret = offset + 1;
                const constant = self.code.items.ptr[ret];
                ret += 1;
                self.constants.items[constant].print(debug);

                const function: *Obj.Function = self.constants.items[constant].as_obj().?.narrow(Obj.Function);
                var j: usize = 0;
                while (j < function.upvalue_count): (j += 1) {
                    const is_local = self.code.items[ret];
                    ret += 1;
                    const index = self.code.items[ret];
                    ret += 1;
                    debug.print("{d:04}    |   {s} {d}\n", .{ret - 2, if (is_local == 1) "local" else "upvalue", index});
                }

                return ret;
            },
            .GetUpvalue => {
                return self.byte_instruction("OP_GET_UPVALUE", offset);
            },
            .SetUpvalue => {
                return self.byte_instruction("OP_SET_UPVALUE", offset);
            },
            .CloseUpvalue => {
                return simple_instruction("OP_CLOSE_UPVALUE", offset);
            },
            .Class => {
                return simple_instruction("OP_CLASS", offset);
            },
            .GetProperty => {
                return self.constant_instruction("OP_GET_PROPERTY", offset);
            },
            .SetProperty => {
                return self.constant_instruction("OP_SET_PROPERTY", offset);
            },
            .Method => {
                return self.constant_instruction("OP_METHOD", offset);
            },
            .Invoke => {
                return self.invoke_instruction("OP_INVOKE", offset);
            }
        }

        return 0;
    }

    fn simple_instruction(name: [:0]const u8, offset: usize) usize {
        debug.print("{s}\n", .{name});
        return offset + 1;
    }

    fn constant_instruction(self: *const Chunk, name: [:0]const u8, offset: usize) usize {
        const constant = self.code.items[offset + 1];
        debug.print("{s} {d} ", .{ name, constant });
        self.constants.items[constant].print(debug);
        debug.print("\n", .{});
        return offset + 2;
    }

    fn byte_instruction(self: *const Chunk, name: [:0]const u8, offset: usize) usize {
        const slot = self.code.items[offset + 1];
        debug.print("{s} {d}\n", .{ name, slot });
        return offset + 2;
    }

    fn jump_instruction(self: *const Chunk, name: [:0]const u8, sign: i32, offset: usize) usize {
        const jump: u16 = (@intCast(u16, self.code.items[offset + 1]) << 8) | self.code.items[offset + 2];
        debug.print("{s} {d} -> {d}\n", .{ name, offset, @intCast(i64, offset) + 3 + sign * jump });
        return offset + 3;
    }


    fn invoke_instruction(self: *const Chunk, name: [:0]const u8, offset: usize) usize {
        const constant = self.code.items[offset + 1];
        const arg_count = self.code.items[offset + 2];
        debug.print("{s} ({d} args) {d} '", .{ name, arg_count, constant });
        debug.print("'\n", .{});
        return offset + 3;
    }
};
